src/v/kafka/server/snc_quota_manager.cc

```diff
@@ -297,10 +298,103 @@ void snc_quota_manager::maybe_arm_balancer_timer() {
+    const auto now = clock::now();
+    const std::vector<borrower_t> borrowers = co_await container().map(
+      [now](snc_quota_manager& qm) -> borrower_t {
+          qm.refill_buckets(now);
+          const inoutpair<quota_t> def = qm.get_deficiency();
+          return {
+            .shard_id = ss::this_shard_id(),
+            .is_borrower = {
+              .in = def.in > 0 ? 1u : 0u, .out = def.out > 0 ? 1u : 0u}};
+      });
+
+    const auto borrowers_count = std::accumulate(
+      borrowers.cbegin(),
+      borrowers.cend(),
+      inoutpair<shard_count_t>{0u, 0u},
+      [](const inoutpair<shard_count_t>& s, const borrower_t& b) {
+          return s + b.is_borrower;
+      });
+    vlog(
+      klog.trace,
+      "qb - Borrowers count: {}, borrowers:",
+      borrowers_count,
+      borrowers);
+    if (is_zero(borrowers_count)) {
+        co_return;
+    }
+
+    // collect quota from lenders
+    const inoutpair<quota_t> collected = co_await container().map_reduce0(
+      [borrowers_count](snc_quota_manager& qm) {
+          const auto surplus_to_collect =
+            [](const quota_t surplus, const shard_count_t borrowers_count) {
+                return borrowers_count > 0 ? surplus / 2 : 0;
+            };
+          const inoutpair<quota_t> surplus = qm.get_surplus();
+          const inoutpair<quota_t> to_collect{
+            .in = surplus_to_collect(surplus.in, borrowers_count.in),
+            .out = surplus_to_collect(surplus.out, borrowers_count.out)};
+          qm.adjust_quota(-to_collect);
+          return to_collect;
+      },
+      inoutpair<quota_t>{0, 0},
+      std::plus{});
+    vlog(klog.trace, "qb - Collected: {}", collected);
+
+    // dispense the collected amount among the borrowers
+    using split_t = decltype(std::div(quota_t{}, shard_count_t{}));
+    inoutpair<split_t> split{{0, 0}, {0, 0}};
+    if (borrowers_count.in > 0) {
+        split.in = std::ldiv(collected.in, borrowers_count.in);
+    }
+    if (borrowers_count.out > 0) {
+        split.out = std::ldiv(collected.out, borrowers_count.out);
+    }
+
+    const auto equal_share_or_0 =
+      [](split_t& split, const shard_count_t is_borrower) -> quota_t {
+        if (is_borrower == 0) {
+            return 0;
+        }
+        if (split.rem == 0) {
+            return split.quot;
+        }
+        --split.rem;
+        return split.quot + 1;
+    };
+    for (const borrower_t& b : borrowers) {
+        const inoutpair<quota_t> share{
+          .in = equal_share_or_0(split.in, b.is_borrower.in),
+          .out = equal_share_or_0(split.out, b.is_borrower.out)};
+        if (!is_zero(share)) {
+            // TBD: make the invocation parallel
+            co_await container().invoke_on(
+              b.shard_id,
+              [share](snc_quota_manager& qm) { qm.adjust_quota(share); });
+        }
+    }
+    vassert(
+      split.in.rem == 0 && split.out.rem == 0,
```

Comment:
why is this guaranteed to be true? the consequences are a hard-stop of the system. is there a less severe way to handle this scenario?

Reply:
I think I will spill the ERROR into the log instead.

Reply:
sounds good